File: xaiop-sdk/java/src/main/java/io/xaiop/stream/Materialize.java
package io.xaiop.stream;

import io.xaiop.Json;
import io.xaiop.XaiopFragment;

/**
 * Materializes parser output into a JSON-facing snapshot value, faithful port of the Node.js
 * SDK's {@code stream/materialize.js}.
 *
 * <p>Root fragments ({@link XaiopFragment}) become their entries object for Snapshot / Diff
 * surfaces (PROT-STREAM JSON-facing). The wire fragment notation stays available through
 * {@link io.xaiop.Parse} when not streaming.
 */
public final class Materialize {
  private Materialize() {}

  /** Deep-cloned JSON snapshot (safe to retain / mutate independently of the parser). */
  public static Object materializeSnapshot(Object parsed) {
    if (parsed instanceof XaiopFragment fragment) {
      return Json.deepClone(fragment.getEntries());
    }
    return Json.deepClone(parsed);
  }

  /**
   * Transfers parser output into a JSON-facing value <b>without</b> cloning a plain document
   * root (ownership moves to the caller). Fragments still deep-clone their entries because the
   * live parser may retain nested references.
   *
   * <p>Use only when the parse result will not be reused by the parser.
   */
  public static Object materializeOwned(Object parsed) {
    if (parsed instanceof XaiopFragment fragment) {
      return Json.deepClone(fragment.getEntries());
    }
    return parsed;
  }
}

File: xaiop-sdk/nodejs/src/core/materialize.ts
// @ts-nocheck
/**
 * Materialize parser output into a JSON-facing snapshot value.
 *
 * Root fragments (`XaiopFragment`) become their `entries` object for Snapshot/Diff
 * surfaces (PROT-STREAM JSON-facing). The wire fragment notation remains available
 * via engine/static parse when not streaming.
 */

import { XaiopFragment } from "./parse.js";
import { cloneJson } from "./clone.js";

/**
 * Deep-cloned JSON snapshot (safe to retain / mutate independently of the parser).
 * @param {unknown} parsed
 * @returns {unknown}
 */
export function materializeSnapshot(parsed) {
  if (parsed instanceof XaiopFragment) {
    return cloneJson(parsed.entries);
  }
  return cloneJson(parsed);
}

/**
 * Transfer parser output into a JSON-facing value **without** cloning a plain
 * document root (ownership moves to the caller). Fragments still deep-clone
 * `entries` because the live parser may retain nested references.
 *
 * Use only when the parse result will not be reused by the parser.
 * @param {unknown} parsed
 * @returns {unknown}
 */
export function materializeOwned(parsed) {
  if (parsed instanceof XaiopFragment) {
    return cloneJson(parsed.entries);
  }
  return parsed;
}
